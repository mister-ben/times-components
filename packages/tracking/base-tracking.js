import React, { Component } from "react";
import hoistNonReactStatic from "hoist-non-react-statics";
import PropTypes from "prop-types";
import resolveAttrs from "./resolve-attrs";
import getDisplayName from "./get-display-name";

export const trackingContextTypes = {
  tracking: PropTypes.shape({
    analytics: PropTypes.func
  })
};

export const makeTracking = (
  WrappedComponent,
  {
    analyticsEvents = [],
    trackingName,
    attrs = {},
    trackView,
    trackChildViews = {}
  } = {}
) => {
  const componentName = getDisplayName(WrappedComponent, trackingName);

  class WithTracking extends Component {
    constructor(props) {
      super(props);

      this.wrappedFuncs = new Map();
      this.childData = {};
      this.viewed = new Set();
    }

    componentDidMount() {
      if (!this.context || !this.context.tracking) {
        return;
      }

      if (trackView) {
        this.context.tracking.analytics({
          object: componentName,
          action: "Rendered",
          attrs: {
            ...resolveAttrs(attrs, this.props)
          }
        });
      }
    }

    onChildView(id, childProps) {
      this.context.tracking.analytics({
        object: `${componentName}Child`,
        action: "Viewed",
        attrs: {
          id,
          ...resolveAttrs(trackChildViews.attrs, childProps)
        }
      });
    }

    getWrappedAnalyticsEvents() {
      return this.wrapWithTracking(
        analyticsEvents,
        (resolvedAttrs, actionName) =>
          this.context.tracking &&
          this.context.tracking.analytics({
            object: componentName,
            action: "Action",
            attrs: {
              actionName,
              ...resolvedAttrs
            }
          })
      );
    }

    wrapWithTracking(eventNames, tracking) {
      return eventNames.reduce((wrappedFuncProps, funcName) => {
        const [origFunc, wrappedFunc] = this.wrappedFuncs.get(funcName) || [];

        if (wrappedFunc && this.props[funcName] === origFunc) {
          return {
            ...wrappedFuncProps,
            [funcName]: wrappedFunc
          };
        }

        const funcWrapped = (...args) => {
          tracking(resolveAttrs(attrs, this.props), funcName);
          return this.props[funcName] && this.props[funcName](...args);
        };

        this.wrappedFuncs.set(funcName, [this.props[funcName], funcWrapped]);

        return {
          ...wrappedFuncProps,
          [funcName]: funcWrapped
        };
      }, {});
    }

    render(customProps = {}) {
      const wrappedProps = {
        ...this.props,
        ...this.getWrappedAnalyticsEvents()
      };

      return <WrappedComponent {...wrappedProps} {...customProps} />;
    }
  }

  WithTracking.displayName = `WithTracking(${componentName})`;

  hoistNonReactStatic(WithTracking, WrappedComponent);

  WithTracking.contextTypes = trackingContextTypes;
  WithTracking.propTypes = WrappedComponent.propTypes;
  WithTracking.defaultProps = WrappedComponent.defaultProps;

  return WithTracking;
};
