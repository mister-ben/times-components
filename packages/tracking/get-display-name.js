const getDisplayName = (WrappedComponent, alternativeName) =>
  alternativeName ||
  WrappedComponent.displayName ||
  WrappedComponent.name ||
  "Component";

export default getDisplayName;
