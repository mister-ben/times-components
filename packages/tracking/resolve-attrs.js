export default (attrs = () => ({}), props = {}, eventArgs = []) =>
  typeof attrs === "function" ? attrs(props, eventArgs) : {};
