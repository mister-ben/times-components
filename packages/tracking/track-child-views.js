import React, { Component } from "react";
import getDisplayName from "react-display-name";
import hoistNonReactStatic from "hoist-non-react-statics";
import trackingContextTypes from "./tracking-context-types";
import resolveAttrs from "./resolve-attrs";

export default (
  WrappedComponent,
  {
    trackingName,
    actionName = "Rendered",
    childIdPropKey,
    getChildren = () => [],
    getAttrs = () => ({})
  } = {}
) => {
  const componentName = getDisplayName(WrappedComponent);

  class WithTrackChildView extends Component {
    constructor(props, context) {
      super(props, context);
      this.viewed = new Set();
      this.handleChildViewed = this.handleChildViewed.bind(this);
    }

    onChildView(id, childProps) {
      this.context.tracking.analytics({
        component: `${trackingName || componentName}Child`,
        action: actionName,
        attrs: {
          ...resolveAttrs(getAttrs, childProps)
        }
      });
    }

    handleChildViewed(childData) {
      const id = childData[childIdPropKey];

      if (this.viewed.has(id)) {
        return;
      }

      this.viewed.add(id);

      const childList = getChildren(this.props);
      const index = childList.findIndex(item => item[childIdPropKey] === id);

      this.onChildView(id, { ...childData, total: childList.length, index });
    }

    render() {
      return (
        <WrappedComponent {...this.props} onViewed={this.handleChildViewed} />
      );
    }
  }

  WithTrackChildView.contextTypes = trackingContextTypes;
  WithTrackChildView.displayName = `WithTrackChildView(${componentName})`;
  WithTrackChildView.propTypes = WrappedComponent.propTypes;
  WithTrackChildView.defaultProps = WrappedComponent.defaultProps;
  hoistNonReactStatic(WithTrackChildView, WrappedComponent);

  return WithTrackChildView;
};
