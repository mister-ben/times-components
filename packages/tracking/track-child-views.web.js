/* eslint-env browser */
import React, { Component } from "react";
import getDisplayName from "react-display-name";
import hoistNonReactStatic from "hoist-non-react-statics";
import trackingContextTypes from "./tracking-context-types";
import resolveAttrs from "./resolve-attrs";

export default (
  WrappedComponent,
  {
    trackingName,
    actionName = "Rendered",
    childIdPropKey,
    getChildren = () => [],
    getAttrs = () => ({})
  } = {}
) => {
  const componentName = getDisplayName(WrappedComponent);

  class WithTrackChildView extends Component {
    constructor(props, context) {
      super(props, context);

      this.childData = {};
      this.viewed = new Set();

      this.observer = new window.IntersectionObserver(
        this.onObserved.bind(this),
        {
          root: null,
          rootMargin: "0px",
          threshold: 1.0
        }
      );
    }

    componentDidMount() {
      this.observeChildren();
    }

    componentDidUpdate() {
      this.observeChildren();
    }

    componentWillUnmount() {
      if (this.observer) {
        this.observer.disconnect();
      }
    }

    onObserved(observed = []) {
      if (!this.context || !this.context.tracking) {
        return;
      }

      observed.forEach(({ intersectionRatio, isIntersecting, target }) => {
        if (
          isIntersecting &&
          intersectionRatio === 1 &&
          !this.viewed.has(target.id)
        ) {
          this.viewed.add(target.id);

          this.onChildView(target.id, this.childData[target.id]);
        }
      });
    }

    onChildView(id, childProps) {
      this.context.tracking.analytics({
        component: `${trackingName || componentName}Child`,
        action: actionName,
        attrs: {
          ...resolveAttrs(getAttrs, childProps)
        }
      });
    }

    observeChildren() {
      const childList = getChildren(this.props);

      childList.forEach((props, index) => {
        if (!this.childData[props[childIdPropKey]]) {
          this.observeChild({
            ...props,
            index,
            total: childList.length
          });
        }
      });
    }

    observeChild(props) {
      this.observer.observe(document.getElementById(props[childIdPropKey]));
      this.childData[props[childIdPropKey]] = props;
    }

    render() {
      return <WrappedComponent {...this.props} />;
    }
  }

  WithTrackChildView.contextTypes = trackingContextTypes;
  WithTrackChildView.displayName = `WithTrackChildView(${componentName})`;
  WithTrackChildView.propTypes = WrappedComponent.propTypes;
  WithTrackChildView.defaultProps = WrappedComponent.defaultProps;
  hoistNonReactStatic(WithTrackChildView, WrappedComponent);

  return WithTrackChildView;
};
