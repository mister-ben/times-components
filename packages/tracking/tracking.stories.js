import React from "react";
import PropTypes from "prop-types";
import {
  Button,
  StyleSheet,
  View,
  Text,
  Platform,
  FlatList
} from "react-native";
// eslint-disable-next-line import/no-extraneous-dependencies
import { action } from "@storybook/addon-actions";
// eslint-disable-next-line import/no-extraneous-dependencies
import { storiesOf } from "@storybook/react-native";
import {
  withTrackRender,
  withTrackingContext,
  withTrackEvents,
  withTrackChildViews
} from "./tracking";

const storybookReporter = action("analytics-event");

const styles = StyleSheet.create({
  box: {
    borderWidth: 0,
    borderColor: "black",
    borderStyle: "solid",
    height: 200,
    width: 200,
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "center"
  }
});
const Box = props => (
  <View id={props.id} style={[styles.box, { backgroundColor: props.color }]}>
    {props.children}
  </View>
);
Box.propTypes = {
  id: PropTypes.string.isRequired,
  children: PropTypes.element,
  color: PropTypes.string.isRequired
};
Box.defaultProps = { children: null };

const BoxWithButtons = props => (
  <View style={[styles.box, { backgroundColor: props.color }]}>
    <Button onPress={() => props.onPress("button 1")} title="Press me" />
    <Button
      color="limegreen"
      onPress={() => props.onPress("button 2")}
      title="Or me"
    />
  </View>
);
BoxWithButtons.propTypes = {
  color: PropTypes.string.isRequired,
  onPress: PropTypes.func.isRequired
};

class Boxes extends React.Component {
  static get propTypes() {
    return {
      onViewed: PropTypes.func.isRequired,
      boxes: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.string.isRequired,
          color: PropTypes.string.isRequired
        })
      ).isRequired
    };
  }
  constructor(props, context) {
    super(props, context);
    this.renderItem = this.renderItem.bind(this);
  }
  renderItem({ item }) {
    this.props.onViewed(item);
    return (
      <Box key={item.id} {...item}>
        <Text style={{ color: "white" }}>{item.id}</Text>
      </Box>
    );
  }
  render() {
    return Platform.OS === "web" ? (
      <View>{this.props.boxes.map(box => this.renderItem({ item: box }))}</View>
    ) : (
      <FlatList
        data={this.props.boxes}
        renderItem={this.renderItem}
        keyExtractor={item => item.id}
      />
    );
  }
}

storiesOf("Tracking", module)
  .add("Render tracking", () => {
    const BoxWithTrackingContext = withTrackingContext(
      withTrackRender(Box, {
        trackingName: "ColoredBox",
        getAttrs: props => ({ color: props.color })
      }),
      { trackingObject: "Story" }
    );

    return (
      <BoxWithTrackingContext
        id="1"
        analyticsStream={storybookReporter}
        color="red"
      />
    );
  })
  .add("Event tracking", () => {
    const BoxWithPressTrackingContext = withTrackingContext(
      withTrackEvents(BoxWithButtons, {
        analyticsEvents: [
          {
            eventName: "onPress",
            actionName: "Pressed",
            trackingName: "ColoredBox",
            getAttrs: (props, eventArgs) => ({ button: eventArgs[0] })
          }
        ]
      }),
      { trackingObject: "Story" }
    );

    return (
      <BoxWithPressTrackingContext
        analyticsStream={storybookReporter}
        color="red"
        onPress={() => {}}
      />
    );
  })
  .add("Child view tracking", () => {
    const boxes = [...Array(50).keys()].map(i => ({
      id: `box-${i + 1}`,
      color: i % 2 === 0 ? "green" : "blue"
    }));
    const BoxesWithTrackingContext = withTrackingContext(
      withTrackChildViews(Boxes, {
        childIdPropKey: "id",
        getChildren: props => props.boxes,
        getAttrs: ({ id, index }) => ({ id, index })
      }),
      { trackingObject: "Story" }
    );
    return (
      <BoxesWithTrackingContext
        onViewed={() => {}}
        boxes={boxes}
        analyticsStream={storybookReporter}
      />
    );
  })
  .add("Scroll depth tracking", () => {
    const boxes = [...Array(50).keys()].map(i => ({
      id: `box-${i + 1}`,
      color: i % 2 === 0 ? "green" : "blue"
    }));
    const BoxesWithTrackingContext = withTrackingContext(
      withTrackChildViews(Boxes, {
        childIdPropKey: "id",
        actionName: "Scrolled",
        getChildren: props => props.boxes,
        getAttrs: props => ({
          scrollDepth: Math.round((props.index + 1) / props.total * 100, 2)
        })
      }),
      { trackingObject: "Story" }
    );
    return (
      <BoxesWithTrackingContext
        onViewed={() => {}}
        boxes={boxes}
        analyticsStream={storybookReporter}
      />
    );
  });
