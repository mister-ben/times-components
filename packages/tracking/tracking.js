import { makeTracking } from "./base-tracking";

export { default as createConsoleReporter } from "./reporters/console-logger";
export { default as tealiumTransformer } from "./transformers/tealium";
export { default as addTrackingContext } from "./tracking-context";

export const addTracking = (
  WrappedComponent,
  {
    analyticsEvents = [],
    trackingName,
    attrs = {},
    trackView,
    trackChildViews = {}
  } = {}
) =>
  class WithTracking extends makeTracking(WrappedComponent, {
    analyticsEvents,
    trackingName,
    attrs,
    trackView,
    trackChildViews
  }) {
    handleChildViewed(data) {
      const id = data[trackChildViews.id];

      if (this.viewed.has(id)) {
        return;
      }

      this.viewed.add(id);

      this.onChildView(id, data, new Date().getTime());
    }

    render() {
      const customProps = trackChildViews.listPath
        ? { onViewed: this.handleChildViewed.bind(this) }
        : {};
      return super.render(customProps);
    }
  };
