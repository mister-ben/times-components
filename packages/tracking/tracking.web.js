/* eslint-env browser */

import _get from "lodash.get";
import { makeTracking } from "./base-tracking";

export { default as createConsoleReporter } from "./reporters/console-logger";
export { default as tealiumTransformer } from "./transformers/tealium";
export { default as addTrackingContext } from "./tracking-context";

const makeDateFromElapsedTime = time =>
  new Date(window.performance.timing.navigationStart + time);

export const addTracking = (
  WrappedComponent,
  {
    analyticsEvents = [],
    trackingName,
    getAttrs = () => ({}),
    trackView,
    trackChildViews = {}
  } = {}
) =>
  class WithTracking extends makeTracking(WrappedComponent, {
    analyticsEvents,
    trackingName,
    getAttrs,
    trackView,
    trackChildViews
  }) {
    constructor(props) {
      super(props);

      if (trackChildViews) {
        const options = {
          root: null,
          rootMargin: "0px",
          threshold: 1.0
        };

        this.observer = new window.IntersectionObserver(
          this.onObserved.bind(this),
          options
        );
      }
    }

    onObserved([{ intersectionRatio, isIntersecting, time, target }]) {
      if (this.context && !this.context.tracking) {
        return;
      }

      if (
        isIntersecting &&
        intersectionRatio === 1 &&
        !this.viewed.has(target.id)
      ) {
        this.viewed.add(target.id);

        this.onChildView(
          target.id,
          this.childData[target.id],
          makeDateFromElapsedTime(time)
        );
      }
    }

    observeChild(props) {
      this.observer.observe(document.getElementById(props[trackChildViews.id]));
      this.childData[props[trackChildViews.id]] = props;
    }

    observeChildren() {
      if (trackChildViews && trackChildViews.listPath) {
        const list = _get(this.props, trackChildViews.listPath, []);
        list.forEach((props, index) => {
          if (!this.childData[props[trackChildViews.id]]) {
            this.observeChild({
              ...props,
              index,
              total: list.length
            });
          }
        });
      }
    }

    componentDidMount() {
      super.componentDidMount();
      this.observeChildren();
    }

    componentDidUpdate() {
      this.observeChildren();
    }

    componentWillUnmount() {
      if (this.observer) {
        this.observer.disconnect();
      }
    }
  };
