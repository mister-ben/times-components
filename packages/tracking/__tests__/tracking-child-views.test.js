import { Text, FlatList } from "react-native";
import React from "react";
import PropTypes from "prop-types";
import renderer from "react-test-renderer";
import { withTrackChildViews } from "../tracking";
import withTestContext from "./test-tracking-context";
import sharedTrackingTests from "./shared-tracking-tests";

class ListComponent extends React.Component {
  static get propTypes() {
    return {
      onViewed: PropTypes.func.isRequired,
      items: PropTypes.arrayOf(
        PropTypes.shape({
          someKey: PropTypes.string,
          someValue: PropTypes.string
        })
      )
    };
  }
  static get defaultProps() {
    return { items: [{ someKey: "1", someValue: "one" }] };
  }
  static get someStatic() {
    return { foo: "bar" };
  }
  constructor(props, context) {
    super(props, context);
    this.renderItem = this.renderItem.bind(this);
  }
  renderItem({ item }) {
    this.props.onViewed(item);
    return <Text>Item {item.someValue}</Text>;
  }
  render() {
    return (
      <FlatList
        data={this.props.items}
        renderItem={this.renderItem}
        keyExtractor={({ someKey }) => someKey}
      />
    );
  }
}

describe("WithTrackChildViews", () => {
  it("tracks child views", () => {
    const reporter = jest.fn();
    const ListWithChildTracking = withTestContext(
      withTrackChildViews(ListComponent, {
        childIdPropKey: "someKey",
        getChildren: props => props.items
      }),
      { trackingObject: "TestObject" }
    );

    renderer.create(
      <ListWithChildTracking
        analyticsStream={reporter}
        onViewed={() => {}}
        items={[
          { someKey: "1", someValue: "one" },
          { someKey: "2", someValue: "two" },
          { someKey: "3", someValue: "three" }
        ]}
      />
    );

    expect(reporter.mock.calls).toMatchSnapshot();
  });

  it("accepts component name override", () => {
    const reporter = jest.fn();
    const ListWithChildTracking = withTestContext(
      withTrackChildViews(ListComponent, {
        childIdPropKey: "someKey",
        getChildren: props => props.items,
        trackingName: "SomeItem"
      })
    );

    renderer.create(
      <ListWithChildTracking
        onViewed={() => {}}
        analyticsStream={reporter}
        items={[{ someKey: "1", someValue: "one" }]}
      />
    );

    expect(reporter).toHaveBeenCalledWith(
      expect.objectContaining({
        component: "SomeItemChild"
      })
    );
  });

  it("accepts action name override", () => {
    const reporter = jest.fn();
    const ListWithChildTracking = withTestContext(
      withTrackChildViews(ListComponent, {
        childIdPropKey: "someKey",
        getChildren: props => props.items,
        actionName: "Scrolled"
      })
    );

    renderer.create(
      <ListWithChildTracking
        onViewed={() => {}}
        analyticsStream={reporter}
        items={[{ someKey: "1", someValue: "one" }]}
      />
    );

    expect(reporter).toHaveBeenCalledWith(
      expect.objectContaining({
        action: "Scrolled"
      })
    );
  });

  it("applies tracking attrs", () => {
    const reporter = jest.fn();
    const ListWithChildTracking = withTestContext(
      withTrackChildViews(ListComponent, {
        childIdPropKey: "someKey",
        getChildren: props => props.items,
        getAttrs: props => ({
          id: props.someKey,
          index: props.index,
          total: props.total
        })
      })
    );

    renderer.create(
      <ListWithChildTracking
        onViewed={() => {}}
        analyticsStream={reporter}
        items={[{ someKey: "1", someValue: "one" }]}
      />
    );

    expect(reporter).toHaveBeenCalledWith(
      expect.objectContaining({
        attrs: expect.objectContaining({ id: "1", index: 0, total: 1 })
      })
    );
  });

  it("can track scroll events", () => {
    const reporter = jest.fn();
    const ListWithChildTracking = withTestContext(
      withTrackChildViews(ListComponent, {
        childIdPropKey: "someKey",
        actionName: "Scrolled",
        getChildren: props => props.items,
        getAttrs: props => ({ depth: (props.index + 1) / props.total * 100 })
      })
    );

    renderer.create(
      <ListWithChildTracking
        onViewed={() => {}}
        analyticsStream={reporter}
        items={[
          { someKey: "1", someValue: "one" },
          { someKey: "2", someValue: "two" }
        ]}
      />
    );

    expect(reporter).toHaveBeenCalledWith(
      expect.objectContaining({
        action: "Scrolled",
        attrs: expect.objectContaining({ depth: 50 })
      })
    );
  });

  sharedTrackingTests(withTrackChildViews);
});
