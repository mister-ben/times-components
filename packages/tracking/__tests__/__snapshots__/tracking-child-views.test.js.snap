// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithTrackChildViews forwards props to wrapped component 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  disabled={false}
  ellipsizeMode="tail"
>
  bar
</Text>
`;

exports[`WithTrackChildViews renders when tracking context is missing 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  disabled={false}
  ellipsizeMode="tail"
>
  foo
</Text>
`;

exports[`WithTrackChildViews tracks child views 1`] = `
Array [
  Array [
    Object {
      "action": "Rendered",
      "attrs": Object {},
      "component": "ListComponentChild",
    },
  ],
  Array [
    Object {
      "action": "Rendered",
      "attrs": Object {},
      "component": "ListComponentChild",
    },
  ],
  Array [
    Object {
      "action": "Rendered",
      "attrs": Object {},
      "component": "ListComponentChild",
    },
  ],
]
`;
